teps/0144-param-enum.md

```diff
@@ -0,0 +1,314 @@
+      - "v1.21"
+      - "v1.20"
+      - "v1.19"
+    default: "v1.21"
+  steps:
+  - name: build
+    image: docker.io/library/golang:$(params.task-version)
+    ...
+```
+
+#### TaskRuns with TaskRef or Remote Resolution
+`Task` users can provide a value for the parameter at the execution time. If the value is in the `enum` list specified by the author, the task will be executed normally. The task will fail the validation with reason `InvalidParamValue` otherwise. 
+
+As an example, the `TaskRun` references the above `Task`. The execution of the `TaskRun` should fail with reason `InvalidParamValue` and a detailed error message since `latest` is NOT in the `enum` list. 
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+  name: run-golang-build
+spec:
+  params:
+  - name: version
+    value: "latest" # 'latest' is not in the enum list
+  taskRef:
+    name: golang-build
+    ...
+```
+
+#### TaskRuns with Embedded TaskSpec
+`Task` users are allowed to specify `enum` for `TaskRuns` with Embedded `TaskSpec` and Tekton will perform the same validation as described above.
+
+``` yaml
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+  generateName: golang-build-run
+spec:
+  params:
+    - name: version
+      value: "v1.21"
+  taskSpec:
+    params:
+      - name: version
+        type: string 
+        enum:
+          - "v1.21"
+          - "v1.20"
+          - "v1.19"
+    ...
+```
+
+While we haven't identified any `enum` use case in this scenario at the time writing this TEP, this is treated as a valid `TaskRun` for `TaskSpec` API compatibility reason.
+
+### Pipelines
+`Pipeline` author can also define a `string` `param` with the `enum` keyword in the `PipelineSpec` `Param` and pass this `string` `param` to the referenced `Task`.
+
+> :warning: Tekton validates the **intersection** of enum set specified in a PipelineTask and the enum set specified in the Pipeline.
+
+#### PipelineTasks with TaskRef or Remote Solution
+The pipeline example references the above `golang-build` Task. The `pipeline-revision` `param` is passed as the value of `task-version` `param` in the `PipelineTask`. In this case, the `Pipeline` user can only pass in the versions specified in the `golang-build` `Tasks` `enum` list (`v1.21`, `v1.20` and `v1.19`) to run the `pipeline` successfully.
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Pipeline
+metadata:
+  name: build-pipeline
+spec:
+  params:
+    - name: pipeline-revision
+      description: revision used for builder image
+      type: string
+  tasks:
+    - name: build-task
+      params:
+        - name: task-version
+          value: $(params.pipeline-revision)
+      taskRef:
+        name: golang-build
+```
+
+The `Pipeline` author can specify `enum` in `spec.params` to put extra restrictions on top of the `enum` specified in the referenced `Task`. Tekton will validate the **intersection** of `enum` specified in the referenced `Task` and the `enum` specified in the `Pipeline` `spec.params`.
```

Comment:
What is the use case for such specification?

Reply:
I think supporting Param Enum at Pipeline level is an useful feature in general.

- With embedded Tasks, the use cases for a `Task` can also be applied to a `Pipeline`.
- With remote/referenced Tasks, it provides more customization capability for the `Pipeline` author, as `Task` is a more reusable and generic unit, which is usually NOT customized for a specific `Pipeline` usage. For example:
  - As a pipeline using [gcs-create-bucket](https://github.com/tektoncd/catalog/tree/main/task/gcs-create-bucket/0.1), I want to further restrict the `StorageClass` to be `ARCHIVE` in this particular pipeline.
  - With an in-organization golang-build Task allowing X approved `golang` images, a pipeline using the task want to further restrict to a subset of X images, based on the version requirement from the feature/code to be built/tested.